File: literallytoast/src/main/java/com/dvoiss/literallytoast/ToastSpringListener.java
package com.dvoiss.literallytoast;

import android.os.Handler;
import android.support.annotation.NonNull;
import android.view.View;
import android.view.WindowManager;
import com.facebook.rebound.SimpleSpringListener;
import com.facebook.rebound.Spring;
import com.facebook.rebound.SpringUtil;

class ToastSpringListener extends SimpleSpringListener {

    @NonNull private final View view;
    @NonNull private final WindowManager windowManager;
    @NonNull private final Handler handler;
    private final long duration;

    ToastSpringListener(@NonNull View view, @NonNull WindowManager windowManager,
        @NonNull Handler handler, long duration) {
        this.view = view;
        this.windowManager = windowManager;
        this.handler = handler;
        this.duration = duration;
    }

    @Override
    public void onSpringUpdate(@NonNull Spring spring) {
        float value = (float) spring.getCurrentValue();
        float barPosition =
            (float) SpringUtil.mapValueFromRangeToRange(value, 0, 1, view.getHeight(), 0);
        view.setTranslationY(barPosition);
    }

    @Override
    public void onSpringAtRest(@NonNull final Spring spring) {
        if (spring.getEndValue() == 1) {
            handler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    spring.setEndValue(0);
                }
            }, duration);
        } else if (spring.getEndValue() <= 0.01) {
            view.setVisibility(View.GONE);
            windowManager.removeView(view);
        }
    }
}

File: literallytoast/src/main/java/com/dvoiss/literallytoast/ToastSpringListener.kt
package com.dvoiss.literallytoast

import android.os.Handler
import android.view.View
import android.view.WindowManager
import com.facebook.rebound.SimpleSpringListener
import com.facebook.rebound.Spring
import com.facebook.rebound.SpringUtil

internal class ToastSpringListener(private val view: View, private val windowManager: WindowManager,
    private val handler: Handler, private val duration: Long) : SimpleSpringListener() {

  /**
   * Update the y position of the toast as the spring updates.
   */
  override fun onSpringUpdate(spring: Spring) {
    view.translationY = SpringUtil.mapValueFromRangeToRange(spring.currentValue, 0.0, 1.0,
        view.height.toDouble(), 0.0).toFloat()
  }

  /**
   * If the spring is at rest and the end-value is 1 then the toast has animated onto the screen
   * and we went to remove it after a certain time.
   *
   * Otherwise if the end-value is at 0 we want to remove it right away.
   */
  override fun onSpringAtRest(spring: Spring) {
    if (spring.endValue == 1.0) {
      handler.postDelayed({ spring.endValue = 0.0 }, duration)
    } else if (spring.endValue <= 0.01) {
      view.visibility = View.GONE
      windowManager.removeView(view)
    }
  }
}
